pythonscript/godot/_hazmat/internal.pyx

```diff
@@ -1,20 +1,62 @@
 cdef bint __pythonscript_verbose = False
-
+import threading
 
 # /!\ This dict is strictly private /!\
 # It contains class objects that are referenced
 # from Godot without refcounting, so droping an
 # item from there will likely cause a segfault
 cdef dict __exposed_classes_per_module = {}
+__all_loaded_classes = {}
+__exposed_classes_lock = threading.Lock()
 
-
+import sys
 cdef object get_exposed_class(str module_name):
-    return __exposed_classes_per_module.get(module_name)
+    sys.stdout.flush()
+    return __exposed_classes_per_module.get(module_name, (None, None))[0]
 
 
 cdef void set_exposed_class(object cls):
-    __exposed_classes_per_module[cls.__module__] = cls
+    sys.stdout.flush()
+
+    modname = cls.__module__
+    # use a threadlock to avoid data races in case godot loads/unloads scripts in multiple threads
+    with __exposed_classes_lock:
+
+        # we must keep track of reference counts for the module when reloading a script,
+        # godot calls pythonscript_script_init BEFORE pythonscript_script_finish
+        # this happens because Godot can make multiple PluginScript instances for the same resource.
+        if modname in __exposed_classes_per_module:
+            cls, mod_refcount = __exposed_classes_per_module[modname]
```

Comment:
This feels weird to me: we were supposed to register `cls`, but somehow forget about this and increase the refcount of the cls registered for this modname.
It would be better to do:
```python
registered_cls, mod_refcount = __exposed_classes_per_module[modname] 
if registered_cls is not cls:
    raise RuntimeError(f"Module `modname` already had class {rigestered_cls!r} registered")
...
```
(Not sure if we should raise an exception or print an error though...)

Reply:
```if registered_cls is not cls```
this will always be False when reloading, since a new class is in memory from a new module, and thus have different id's. I guess we can do:
```if registered_cls.__name__ != cls.__name__```
and I think we should print an error, following godot's philosophy of being error tolerant